<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Index of the pipeline parameters · TarGene</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/nextflow_params/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">TarGene</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Setting a data source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Confounding Adjustment</a></li><li><a class="tocitem" href="../parameter_specification/">Parameter plans</a></li><li><a class="tocitem" href="../tmle/">Targeted Maximum Likelihood Estimation</a></li><li><a class="tocitem" href="../sieve_variance/">Sieve Variance Plateau Correction</a></li><li class="is-active"><a class="tocitem" href>Index of the pipeline parameters</a><ul class="internal"><li><a class="tocitem" href="#UK-Biobank-Data"><span>UK-Biobank Data</span></a></li><li><a class="tocitem" href="#Statistical-Parameters"><span>Statistical Parameters</span></a></li><li><a class="tocitem" href="#TMLE"><span>TMLE</span></a></li><li><a class="tocitem" href="#Sieve-Variance-Correction"><span>Sieve Variance Correction</span></a></li><li><a class="tocitem" href="#Miscellaneous"><span>Miscellaneous</span></a></li></ul></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../project_organization/">Project Organization</a></li><li><a class="tocitem" href="../contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">User Guide</a></li><li class="is-active"><a href>Index of the pipeline parameters</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Index of the pipeline parameters</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/nextflow_params.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-the-pipeline-parameters"><a class="docs-heading-anchor" href="#Index-of-the-pipeline-parameters">Index of the pipeline parameters</a><a id="Index-of-the-pipeline-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-the-pipeline-parameters" title="Permalink"></a></h1><p>Here is a list of all the pipeline parameters:</p><h2 id="UK-Biobank-Data"><a class="docs-heading-anchor" href="#UK-Biobank-Data">UK-Biobank Data</a><a id="UK-Biobank-Data-1"></a><a class="docs-heading-anchor-permalink" href="#UK-Biobank-Data" title="Permalink"></a></h2><ul><li><code>ENCRYPTED_DATASET</code>: Path to a UK-Biobank encrypted main dataset.</li><li><code>ENCODING_FILE</code>: If an encrypted dataset is provided, an encoding file must accompany it.</li><li><code>DECRYPTED_DATASET</code>: Path to a UK-Biobank decrypted main dataset. If set, <code>ENCRYPTED_DATASET</code> is ignored.</li><li><code>TRAITS_CONFIG</code>: Configuration file describing which traits should be extracted from the main dataset.</li><li><code>WITHDRAWAL_LIST</code>: List of participants withdrawn from the study.</li><li><code>QC_FILE</code>: Genotyping quality control file from the UK-Biobank study.</li><li><code>UKBB_BED_FILES</code>: Path expression to PLINK BED files.</li><li><code>UKBB_BGEN_FILES</code>: Path expression to iputed BGEN files.</li></ul><h2 id="Statistical-Parameters"><a class="docs-heading-anchor" href="#Statistical-Parameters">Statistical Parameters</a><a id="Statistical-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Statistical-Parameters" title="Permalink"></a></h2><ul><li><code>PARAMETER_PLAN</code>: One of &quot;FROM<em>PARAM</em>FILES&quot;, &quot;FROM_ACTORS&quot;. See the section.</li></ul><p>&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp; If <code>PARAMETER_PLAN</code>=&quot;FROM<em>PARAM</em>FILES&quot;:</p><ul><li><code>PARAMETER_FILES</code>: Path expression to the parameter files.</li></ul><p>&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp;&amp;nbsp; If <code>PARAMETER_PLAN</code>=&quot;FROM_ACTORS&quot;:</p><ul><li><code>BQTLS</code>: CSV file containing binding quantitative trait loci.</li><li><code>TRANS_ACTORS</code>: CSV file containing quantitative trait loci potentially interacting with the previous bqtls.</li><li><code>EXTRA_CONFOUNDERS</code>: Path to additional confounders file, one per line, no header.</li><li><code>EXTRA_COVARIATES</code>: Path to additional covariates file, one per line, no header.</li><li><code>ENVIRONMENTALS</code>: Path to additional environmental treatments file, one per line, no header.</li><li><code>ORDERS</code>: Comma separated list describing the order of desired interaction parameters, 1 for the ATE (no interaction), 2 for pairwise interactions etc... e.g. &quot;1,2&quot;</li></ul><h2 id="TMLE"><a class="docs-heading-anchor" href="#TMLE">TMLE</a><a id="TMLE-1"></a><a class="docs-heading-anchor-permalink" href="#TMLE" title="Permalink"></a></h2><ul><li><code>ESTIMATORFILE</code>: YAML configuration file describing the nuisance parameters learners.</li><li><code>PHENOTYPES_BATCH_SIZE</code>: For a given causal model, phenotypes can be batched to save computational time. Setting this value to 0 will result in max size batches.</li></ul><h2 id="Sieve-Variance-Correction"><a class="docs-heading-anchor" href="#Sieve-Variance-Correction">Sieve Variance Correction</a><a id="Sieve-Variance-Correction-1"></a><a class="docs-heading-anchor-permalink" href="#Sieve-Variance-Correction" title="Permalink"></a></h2><ul><li><code>GRM_NSPLITS</code>: To fasten GRM computation, it is typically split in batches.</li><li><code>NB_VAR_ESTIMATORS</code>: Number of sieve variance estimates per curve. Setting this value to 0 results in skipping sieve variance correction.</li><li><code>MAX_TAU</code>: Variance estimates are computed for tau ranging from 0 to MAX_TAU</li><li><code>PVAL_SIEVE</code>: To save computation time and disk, only parameters with a p-value below this threshold are considered for sieve variance correction.</li></ul><h2 id="Miscellaneous"><a class="docs-heading-anchor" href="#Miscellaneous">Miscellaneous</a><a id="Miscellaneous-1"></a><a class="docs-heading-anchor-permalink" href="#Miscellaneous" title="Permalink"></a></h2><ul><li><code>CALL_THRESHOLD</code>: Variants from BGEN files are called beyond that threhsold.</li><li><code>POSITIVITY_CONSTRAINT</code>: Treatment variables rarest configuration should have at least that frequency.</li><li><code>MAF_THRESHOLD</code>: Only variants with that minor allele frequency are considered</li><li><code>NB_PCS</code>: Number of principal components to include in confouding adjustment</li><li><code>FLASHPCA_EXCLUSION_REGIONS</code>: Special file from <a href="https://github.com/gabraham/flashpca">flashpca</a> repository.</li><li><code>OUTDIR</code>: Output directory</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../sieve_variance/">« Sieve Variance Plateau Correction</a><a class="docs-footer-nextpage" href="../runtime_considerations/">Some Runtime considerations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Wednesday 18 January 2023 10:12">Wednesday 18 January 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
