<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>The Randomized Variants Workflow · TarGene</title><meta name="title" content="The Randomized Variants Workflow · TarGene"/><meta property="og:title" content="The Randomized Variants Workflow · TarGene"/><meta property="twitter:title" content="The Randomized Variants Workflow · TarGene"/><meta name="description" content="Documentation for TarGene."/><meta property="og:description" content="Documentation for TarGene."/><meta property="twitter:description" content="Documentation for TarGene."/><meta property="og:url" content="https://TARGENE.github.io/targene-pipeline/negative_control/randomized_tests/"/><meta property="twitter:url" content="https://TARGENE.github.io/targene-pipeline/negative_control/randomized_tests/"/><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/negative_control/randomized_tests/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">TarGene</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../../overview/">Overview</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">The TarGene Workflow </span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../targene/overview/">The TarGene Workflow</a></li><li><a class="tocitem" href="../../targene/data_sources/">Setting a Data Source</a></li><li><a class="tocitem" href="../../targene/confounding_adjustment/">Adjusting for confounders</a></li><li><a class="tocitem" href="../../targene/study_designs/">Study Designs</a></li><li><a class="tocitem" href="../../targene/tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../../targene/sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../../targene/miscellaneous/">Tweaking additional behaviour</a></li><li><a class="tocitem" href="../../targene/configuration/">Configuration</a></li><li><a class="tocitem" href="../../targene/runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-2-3" type="checkbox" checked/><label class="tocitem" for="menuitem-2-3"><span class="docs-label">Negative Control</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../overview/">Negative Control Overview</a></li><li><a class="tocitem" href="../permutation_tests/">The Permutation Test Workflow</a></li><li class="is-active"><a class="tocitem" href>The Randomized Variants Workflow</a><ul class="internal"><li><a class="tocitem" href="#Example-Run-Command"><span>Example Run Command</span></a></li><li><a class="tocitem" href="#Configuration"><span>Configuration</span></a></li></ul></li></ul></li><li><a class="tocitem" href="../../secondary_workflows/pca/">The PCA Workflow</a></li><li><a class="tocitem" href="../../secondary_workflows/make_dataset/">The Make Dataset Workflow</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../../developper_guide/project_organization/">Project Organization</a></li><li><a class="tocitem" href="../../developper_guide/contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li><a class="is-disabled">Negative Control</a></li><li class="is-active"><a href>The Randomized Variants Workflow</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>The Randomized Variants Workflow</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/negative_control/randomized_tests.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="The-Randomized-Variants-Workflow"><a class="docs-heading-anchor" href="#The-Randomized-Variants-Workflow">The Randomized Variants Workflow</a><a id="The-Randomized-Variants-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#The-Randomized-Variants-Workflow" title="Permalink"></a></h1><p>In some interaction studies the variants are chosen based on a priori knowledge, typically they are QTLs. Consider two variants <span>$V_1$</span> and <span>$V_2$</span>, that have been pre-identified to be eQTLs for a given gene&#39;s transcript. Then it makes sense to run a phenome-wide association study between these two variants because they potentially interact via the gene&#39;s transcript. This can be done via the <a href="../../targene/overview/#The-TarGene-Workflow">TarGene workflow</a> for instance and discoveries can be made. Because these two variants have been carefully chosen, replacing any of them for a random variant whose function is unknown should not result in an interaction. This is the basis of the randomized variants workflow.</p><div class="admonition is-info"><header class="admonition-header">Note</header><div class="admonition-body"><p>By chance, one could pick a random variant also interacting with the other QTL. However, on average we should not deviate from the null hypothesis of no association. This can be checked via examination of the QQ plot generated by TarGene. This is addressed by selecting a certain number of random variants (<code>N_RANDOM_VARIANTS</code>) instead of 1.</p></div></div><h2 id="Example-Run-Command"><a class="docs-heading-anchor" href="#Example-Run-Command">Example Run Command</a><a id="Example-Run-Command-1"></a><a class="docs-heading-anchor-permalink" href="#Example-Run-Command" title="Permalink"></a></h2><pre><code class="language-bash hljs">nextflow run https://github.com/TARGENE/targene-pipeline/ -r TAG -entry RANDOMIZATION_TEST -profile P -resume</code></pre><h2 id="Configuration"><a class="docs-heading-anchor" href="#Configuration">Configuration</a><a id="Configuration-1"></a><a class="docs-heading-anchor-permalink" href="#Configuration" title="Permalink"></a></h2><h3 id="Main-Outputs"><a class="docs-heading-anchor" href="#Main-Outputs">Main Outputs</a><a id="Main-Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Main-Outputs" title="Permalink"></a></h3><p>The workflow will produce the following output in the output directory (<code>OUTDIR</code>, defaults to <code>results</code>): an estimand file named &quot;random<em>variants</em>estimands.jls&quot;. This can be further used by running a classic <a href="../../targene/overview/#The-TarGene-Workflow">TarGene workflow</a> with <code>STUDY_DESIGN=CUSTOM</code>.</p><h3 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h3><ul><li><strong><code>RESULTS_FILE</code> (required)</strong>: Path to an HDF5 results file generated by a previous TarGene run.</li><li><strong><code>VARIANTS_TO_RANDOMIZE</code> (required)</strong>: A text file containing the list of variants (one rsid per line) for which random candidates must be proposed.</li><li><strong><code>BGEN_FILES</code> (required)</strong>: Path to imputed BGEN files from which the variants in <code>VARIANTS_TO_RANDOMIZE</code> will be found.</li></ul><h3 id="Main-Options"><a class="docs-heading-anchor" href="#Main-Options">Main Options</a><a id="Main-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Main-Options" title="Permalink"></a></h3><ul><li><strong><code>N_RANDOM_VARIANTS</code> (optional, default: 10)</strong>: The number of random variants to propose for each variant in <code>VARIANTS_TO_RANDOMIZE</code>.</li><li><strong><code>MAF_MATCHING_RELTOL</code> (optional, default: 0.05)</strong>: Each proposed random variant should have the same minor allele frequency as the origin variant, up to a relative threshold.</li><li><strong><code>POSITIVITY_CONSTRAINT</code> (optional, default: 0.01)</strong>: When the list of estimands is generated or validated. Treatment variables&#39; rarest configuration should have at least that frequency. For example if the treatment variables are two variants with minor allele A and T respectively. The rarest configuration will be (AA, TT) and should have a frequency of at least <code>POSITIVITY_CONSTRAINT</code>.</li><li><strong><code>PVAL_THRESHOLD</code> (optional, default: 0.05)</strong>: Only results with a p-value below this threshold are considered for permutation testing.</li><li><strong><code>ESTIMATOR_KEY</code> (optional, default: TMLE)</strong>: The p-value for <code>PVAL_THRESHOLD</code> is computed using the result from this estimator.</li></ul><h3 id="Secondary-Options"><a class="docs-heading-anchor" href="#Secondary-Options">Secondary Options</a><a id="Secondary-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Secondary-Options" title="Permalink"></a></h3><ul><li><strong><code>RNG</code> (optional, default: 123)</strong>: General random seed used to induce permutation.</li><li><strong><code>VERBOSITY</code> (optional, default: 0)</strong>: Verbosity level of the the Workflow&#39;s processes.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../permutation_tests/">« The Permutation Test Workflow</a><a class="docs-footer-nextpage" href="../../secondary_workflows/pca/">The PCA Workflow »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.3.0 on <span class="colophon-date" title="Monday 18 March 2024 13:49">Monday 18 March 2024</span>. Using Julia version 1.10.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
