<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Describing the causal parameters of interest · TarGene</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/parameter_specification/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">TarGene</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Setting a data source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Adjusting for confounders</a></li><li class="is-active"><a class="tocitem" href>Describing the causal parameters of interest</a><ul class="internal"><li><a class="tocitem" href="#Parameter-Files"><span>Parameter Files</span></a></li><li><a class="tocitem" href="#Parameter-Plans"><span>Parameter Plans</span></a></li><li><a class="tocitem" href="#Parallelization"><span>Parallelization</span></a></li></ul></li><li><a class="tocitem" href="../tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../miscellaneous/">Tweaking additional behaviour</a></li><li><a class="tocitem" href="../negative_control/">Running negative control checks</a></li></ul></li><li><span class="tocitem">Miscellaneous</span><ul><li><a class="tocitem" href="../nextflow_params/">Index of the pipeline parameters</a></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../project_organization/">Project Organization</a></li><li><a class="tocitem" href="../contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">User Guide</a></li><li class="is-active"><a href>Describing the causal parameters of interest</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Describing the causal parameters of interest</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/parameter_specification.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Describing-the-causal-parameters-of-interest"><a class="docs-heading-anchor" href="#Describing-the-causal-parameters-of-interest">Describing the causal parameters of interest</a><a id="Describing-the-causal-parameters-of-interest-1"></a><a class="docs-heading-anchor-permalink" href="#Describing-the-causal-parameters-of-interest" title="Permalink"></a></h1><h2 id="Parameter-Files"><a class="docs-heading-anchor" href="#Parameter-Files">Parameter Files</a><a id="Parameter-Files-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-Files" title="Permalink"></a></h2><p>In this section, by parameter, we mean the statistical parameter that represents the scientific quantity of interest and will be estimated via TarGene. The complete specification of a parameter requires the description of a causal model which can be represented by the following graph.</p><div style="text-align:center">
<img src="../assets/causal_graph.png" alt="Causal Model" style="width:600px;"/>
</div><p>Parameters are specified by a YAML file containing the list of parameters to be estimated. Each parameter is fully determined by four fields:</p><ul><li><code>type</code>: The parameter type, one of:<ul><li>ATE: Average Treatment Effect</li><li>IATE: Interaction Average Treatment Effect</li><li>CM: Conditional Mean</li></ul></li><li><code>target</code>: The trait of interest, as specified in the <code>TRAITS_CONFIG</code> file in the <code>phenotypes/name</code> field. You can also use the wildcard &quot;*&quot; to signify that you want to estimate this parameter accross all traits in the dataset.</li><li><code>treatment</code>: The treatment variables and associated control/case settings (see example below).</li><li><code>confounders</code>: A list of confounding variables. Note that principal components will be added to that list by default and must not be provided here. You can provide an empty list.</li><li><code>covariates</code>: This is optional and correspond to a list of additional covariates for the prediction of the trait.</li></ul><p>Here is an example parameter file:</p><pre><code class="language-yaml hljs">
Parameters:
  - type: IATE
    target: &quot;*&quot;
    treatment: (RSID_10 = (control = &quot;AA&quot;, case = &quot;AC), Sun_Exposure = (control = 1, case = 0))
    confounders: [Economic_background]
    covariates: [Age]
  - type: ATE
    target: PHENOTYPE_2
    treatment: (RSID_10 = (control = &quot;AA&quot;, case = &quot;CC&quot;), RSID_100 = (control = &quot;GC&quot;, case = &quot;CC&quot;))
    confounders: []
  - type: CM
    target: PHENOTYPE_2
    treatment: (RSID_10 = &quot;AA&quot;,)
    confounders: []</code></pre><p>Note that variants can be encoded either with an explicit string representation (e.g. &quot;AC&quot;) or via an integer (0, 1, 2) representing the number of minor alleles in the genotype. All variants in a parameter file should however respect the same encoding. String encoded variants are later one-hot-encoded during the TMLE step.</p><h2 id="Parameter-Plans"><a class="docs-heading-anchor" href="#Parameter-Plans">Parameter Plans</a><a id="Parameter-Plans-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-Plans" title="Permalink"></a></h2><p>At the moment, there are two main ways one can specify parameters that need to be estimated during a targene-pipeline run. This is done via the <code>MODE</code> parameter.</p><h3 id="PARAMETER_PLAN-FROM_PARAM_FILE"><a class="docs-heading-anchor" href="#PARAMETER_PLAN-FROM_PARAM_FILE"><code>PARAMETER_PLAN</code> = <code>FROM_PARAM_FILE</code></a><a id="PARAMETER_PLAN-FROM_PARAM_FILE-1"></a><a class="docs-heading-anchor-permalink" href="#PARAMETER_PLAN-FROM_PARAM_FILE" title="Permalink"></a></h3><p>This is the most general setting and should match the needs of any project, however it requires some preliminary work. In this setting, one typically provides a set of parameter files as described above. If you are interested in only a few parameters it may be acceptable to write them by hand. Otherwise it is best to generate them using a programming language (for instance using <a href="https://targene.github.io/TMLE.jl/stable/">TMLE.jl</a>). The path to those parameters is then provided with the <code>PARAMETER_FILE</code> nextflow parameter. See the previous section on <a href="#Parameter-Files">Parameter Files</a>.</p><h3 id="PARAMETER_PLAN-FROM_ACTORS"><a class="docs-heading-anchor" href="#PARAMETER_PLAN-FROM_ACTORS"><code>PARAMETER_PLAN</code> = <code>FROM_ACTORS</code></a><a id="PARAMETER_PLAN-FROM_ACTORS-1"></a><a class="docs-heading-anchor-permalink" href="#PARAMETER_PLAN-FROM_ACTORS" title="Permalink"></a></h3><p>In this setting the goal is to infer the interaction effect between multiple variants and potential external factors, interacting together via a specific biological mechanism. Typically, multiple sets of variants are of interest and each set is identified with a specific molecule, contributing to the mechanism. In particular, it is assumed that a set of variants, usually binding quantitative trait loci (bQTLs) play a pivotal role. All interactions of interest are thus defined with respect to that set of genetic variations. Let&#39;s Consider the following scenario: we know that a transcription factor binds to molecules <code>x</code> and <code>y</code> and then differentially binds to specific regions in the genome (<code>bQTLs</code>) to regulate downstream genes. We suspect that an alteration of this mechanism is responsible for some diseases. A set of <code>xQTLs</code>, associated with the expression of <code>x</code> and a set of <code>yQTLs</code> associated with the expression of <code>y</code> have been identified. Together <code>xQTLs</code> and <code>yQTLs</code> variants are termed &quot;trans actors&quot;. We further suspect that some environmental factors may influence this process. From that scenario, there are many questions that can be asked, for instance : &quot;What is the interaction effects of a bQTL with an environmental factor?&quot;. This is a simple pairwise interaction setting and more complex scenarios can be envisaged as described in the following graph.</p><div style="text-align:center">
<img src="../assets/from_actors.png" alt="FromActors" style="width:600px;"/>
</div><p>Let us now turn to the pipeline specification for this parameter plan:</p><ul><li><code>BQTLS</code>: A path to a <code>.csv</code> file containing at least an <code>ID</code> column for each rsID and an optional <code>CHR</code> column for the chromosome on which the SNP is located.</li><li><code>TRANS_ACTORS</code>: A path prefix to a set of <code>.csv</code> files identifying different trans-acting variants. Each file has the same format as for the <code>bQTLs</code>.</li><li><code>ENVIRONMENTALS</code>: A path to a <code>.txt</code> file containing a list of environmental exposures with no header and one exposure per line. Each exposure should be available from the trait dataset.</li><li><code>EXTRA_COVARIATES</code>: A path to a <code>.txt</code> file containing a list of extra covariate variables with no header and one variable per line. Each variable should be available from the trait dataset.</li><li><code>EXTRA_CONFOUNDERS</code>: A path to a <code>.txt</code> file containing a list of extra confounding variables with no header and one variable per line. Each variable should be available from the trait dataset.</li><li><code>ORDERS</code>: A comma separated string that specifies the various interaction orders of interest. All combinations satisfying the positivity constraint will be generated. The order 1 corresponds to the Average Treatment Effect (ATE) for <code>bQTLs</code>, any higher order corresponds to the Interaction Average Treatment Effect (IATE) for the various actors. For example, in the previous scenario, assume we provided <code>ORDERS</code>=<code>1,2</code>. This would generate parameter files for the estimation of all:<ul><li>ATEs parameters for all bQTLs</li><li>IATEs parameters for all (bQTLs, xQTLs), (bQTLs, yQTLs), (bQTLs, Envs) pairs.</li></ul></li></ul><h2 id="Parallelization"><a class="docs-heading-anchor" href="#Parallelization">Parallelization</a><a id="Parallelization-1"></a><a class="docs-heading-anchor-permalink" href="#Parallelization" title="Permalink"></a></h2><p>Since the same estimator for <code>p(T|W)</code> can be used for multiple target parameters, it may be useful to batch phenotypes using <code>BATCH_SIZE</code>(default: 400) in order to reduce the computational burden.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../confounding_adjustment/">« Adjusting for confounders</a><a class="docs-footer-nextpage" href="../tmle/">Specifying a Targeted Estimator »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Monday 8 May 2023 17:55">Monday 8 May 2023</span>. Using Julia version 1.8.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
