<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Running negative control checks · TarGene</title><meta name="title" content="Running negative control checks · TarGene"/><meta property="og:title" content="Running negative control checks · TarGene"/><meta property="twitter:title" content="Running negative control checks · TarGene"/><meta name="description" content="Documentation for TarGene."/><meta property="og:description" content="Documentation for TarGene."/><meta property="twitter:description" content="Documentation for TarGene."/><meta property="og:url" content="https://TARGENE.github.io/targene-pipeline/negative_control/"/><meta property="twitter:url" content="https://TARGENE.github.io/targene-pipeline/negative_control/"/><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/negative_control/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">TarGene</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Setting a data source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Adjusting for confounders</a></li><li><a class="tocitem" href="../parameter_specification/">Describing the causal parameters of interest</a></li><li><a class="tocitem" href="../tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../miscellaneous/">Tweaking additional behaviour</a></li><li class="is-active"><a class="tocitem" href>Running negative control checks</a><ul class="internal"><li><a class="tocitem" href="#Permutation-tests"><span>Permutation tests</span></a></li><li><a class="tocitem" href="#Non-functional-randomly-drawn-variants"><span>Non-functional randomly drawn variants</span></a></li></ul></li></ul></li><li><span class="tocitem">Miscellaneous</span><ul><li><a class="tocitem" href="../nextflow_params/">Index of the pipeline parameters</a></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../project_organization/">Project Organization</a></li><li><a class="tocitem" href="../contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">User Guide</a></li><li class="is-active"><a href>Running negative control checks</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Running negative control checks</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/negative_control.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Running-negative-control-checks"><a class="docs-heading-anchor" href="#Running-negative-control-checks">Running negative control checks</a><a id="Running-negative-control-checks-1"></a><a class="docs-heading-anchor-permalink" href="#Running-negative-control-checks" title="Permalink"></a></h1><p>After running TarGene, you may want to perform some quality control checks on your positive hits. Since true positives are not ubiquitous in biology, this is done by negative control checks. The corresponding pipeline can be run by:</p><pre><code class="language-bash hljs">nextflow run https://github.com/TARGENE/targene-pipeline -entry negativeControl -r vX -profile P -resume -with-trace -with-report</code></pre><p>see the <a href="../overview/#Overview">Overview</a> section for more info on the command line options.</p><p>At the moment, there are two types of analyses you can perform with this sub-pipeline: permutation tests and comparisons with &quot;non-functional&quot; random variants drawn from the genome. Furthermore, this is restricted to interaction parameters (IATE). In both cases, you will need to define positive hits. This is traditionally done via p-value filtering, you will thus need to specify both a pvalue column (from the &quot;summary.csv&quot; file) and a threshold: <code>PVAL_COL</code> and <code>PVAL_THRESHOLD</code> parameters.</p><h2 id="Permutation-tests"><a class="docs-heading-anchor" href="#Permutation-tests">Permutation tests</a><a id="Permutation-tests-1"></a><a class="docs-heading-anchor-permalink" href="#Permutation-tests" title="Permalink"></a></h2><p>For each IATE hit, we can perturb the data by independently shuffling one or more columns in the data. By doing so, the interaction will be broken and become &quot;non-significant&quot;. Because there are up to K variables in a IATE parameter, we can also permute up to K+1 variables. For instance, assuming we found a significant interaction between rs1234 and rs5678 on diabetes, we could permute:</p><ul><li>Any one of (rs1234, rs5678, diabetes) leading to 3 order 1 permutation tests.</li><li>Any two of (rs1234, rs5678, diabetes) resulting also in 3 order 2 permutation tests.</li><li>All three of (rs1234, rs5678, diabetes) leading to 1 order 3 permutation test.</li></ul><p>Specifying those orders is made via the <code>PERMUTATION_ORDERS</code> variable which is a comma separated string of orders (e.g. &quot;1,2&quot;).</p><p>The results of those permutation tests will be found in the <code>$(OUTDIR)/permutation_summary.csv</code> file.</p><h2 id="Non-functional-randomly-drawn-variants"><a class="docs-heading-anchor" href="#Non-functional-randomly-drawn-variants">Non-functional randomly drawn variants</a><a id="Non-functional-randomly-drawn-variants-1"></a><a class="docs-heading-anchor-permalink" href="#Non-functional-randomly-drawn-variants" title="Permalink"></a></h2><p>If you are using the <code>PARAMETER_PLAN</code> = <code>FROM_ACTORS</code>, it is likely that you have not chosen the genetic variants at random. Mainly trans-acting variants have been defined because of a likely role in a biological mechanism. If this is the case, replacing a trans-actor by a random variant anywhere on the genome is likely to break the interaction. While unlikely, by chance alone (or lack thereof), one could pick a random trans-actor which is also interacting with any of the bQTLs. In order to account for that it is recommended to instead select a certain number of rancom variants, denoted by <code>N_RANDOM_VARIANTS</code> (default: 10). Furthermore, we enforce two criteria on each of the randomly chosen random variants:</p><ul><li>Its minor allele frequency should match that of the original trans-actor up to <code>MAF_MATCHING_RELTOL</code> (relative tolerance, default: 0.05).</li><li>It shouldn&#39;t lie in a known regulatory region.</li></ul><p>The result of this part of the pipeline is a parameter file located at <code>$(OUTDIR)/random_variants_parameters.yaml</code> directory. To perform the actual tests, you will have to run TarGene again on that parameter file with the  <code>PARAMETER_PLAN</code> = <code>FROM_PARAM_FILE</code> mode.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../miscellaneous/">« Tweaking additional behaviour</a><a class="docs-footer-nextpage" href="../nextflow_params/">Index of the pipeline parameters »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Monday 4 December 2023 11:57">Monday 4 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
