<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Index of the pipeline parameters · TarGene</title><meta name="title" content="Index of the pipeline parameters · TarGene"/><meta property="og:title" content="Index of the pipeline parameters · TarGene"/><meta property="twitter:title" content="Index of the pipeline parameters · TarGene"/><meta name="description" content="Documentation for TarGene."/><meta property="og:description" content="Documentation for TarGene."/><meta property="twitter:description" content="Documentation for TarGene."/><meta property="og:url" content="https://TARGENE.github.io/targene-pipeline/nextflow_params/"/><meta property="twitter:url" content="https://TARGENE.github.io/targene-pipeline/nextflow_params/"/><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/nextflow_params/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">TarGene</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Setting a data source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Adjusting for confounders</a></li><li><a class="tocitem" href="../study_designs/">Study Designs</a></li><li><a class="tocitem" href="../tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../miscellaneous/">Tweaking additional behaviour</a></li><li><a class="tocitem" href="../negative_control/">Running negative control checks</a></li></ul></li><li><span class="tocitem">Miscellaneous</span><ul><li class="is-active"><a class="tocitem" href>Index of the pipeline parameters</a><ul class="internal"><li><a class="tocitem" href="#[Setting-a-data-source](@ref)"><span>Setting a data source</span></a></li><li><a class="tocitem" href="#[Adjusting-for-confounders](@ref)"><span>Adjusting for confounders</span></a></li><li><a class="tocitem" href="#[Study-Designs](@ref)"><span>Study Designs</span></a></li><li><a class="tocitem" href="#[Specifying-a-Targeted-Estimator](@ref)"><span>Specifying a Targeted Estimator</span></a></li><li><a class="tocitem" href="#[Correcting-for-population-relatedness](@ref)"><span>Correcting for population relatedness</span></a></li><li><a class="tocitem" href="#[Tweaking-additional-behaviour](@ref)"><span>Tweaking additional behaviour</span></a></li><li><a class="tocitem" href="#[Running-negative-control-checks](@ref)"><span>Running negative control checks</span></a></li></ul></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../project_organization/">Project Organization</a></li><li><a class="tocitem" href="../contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Miscellaneous</a></li><li class="is-active"><a href>Index of the pipeline parameters</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Index of the pipeline parameters</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/nextflow_params.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-the-pipeline-parameters"><a class="docs-heading-anchor" href="#Index-of-the-pipeline-parameters">Index of the pipeline parameters</a><a id="Index-of-the-pipeline-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-the-pipeline-parameters" title="Permalink"></a></h1><p>Here is a list of all the pipeline parameters:</p><h2 id="[Setting-a-data-source](@ref)"><a class="docs-heading-anchor" href="#[Setting-a-data-source](@ref)"><a href="../data_sources/#Setting-a-data-source">Setting a data source</a></a><a id="[Setting-a-data-source](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Setting-a-data-source](@ref)" title="Permalink"></a></h2><ul><li><strong><code>COHORT</code> (required):</strong> Current default for this is UKBB. If set to a value other than UKBB, this will not run UKBB-specific trait extraction. If <code>COHORT</code> is not UKBB, you must specify your trait data in the <code>TRAITS_DATASET</code> parameter.</li><li><strong><code>TRAITS_DATASET</code> (required):</strong> Path to a traits dataset. If you are running this for a non-UKBB cohort, your sample IDs must be specified in the first column of this CSV file, with the column name <code>SAMPLE_ID</code>.</li><li><strong><code>BED_FILES</code> (required)</strong>: Path expression to PLINK BED files.</li><li><strong><code>BGEN_FILES</code> (required)</strong>: Path expression to imputed BGEN files.</li><li><strong><code>UKB_CONFIG</code> (required for COHORT=UKBB)</strong>: Configuration file describing which traits should be extracted from the main dataset.</li><li><code>UKB_ENCODING_FILE</code> (optional): If the <code>TRAITS_DATASET</code> is encrypted, an encoding file must be provided.</li><li><code>UKB_WITHDRAWAL_LIST</code> (optional): List of participants withdrawn from the study.</li><li><code>QC_FILE</code> (optional): Genotyping quality control file from the UK-Biobank study.</li></ul><h2 id="[Adjusting-for-confounders](@ref)"><a class="docs-heading-anchor" href="#[Adjusting-for-confounders](@ref)"><a href="../confounding_adjustment/#Adjusting-for-confounders">Adjusting for confounders</a></a><a id="[Adjusting-for-confounders](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Adjusting-for-confounders](@ref)" title="Permalink"></a></h2><ul><li><code>LD_BLOCKS</code> (optional): A path to pre-identified linkage disequlibrium blocks around the variants that will be queried for causal effect estimation. Those will be removed from the data. It is good practice to specify <code>LD_BLOCKS</code>, as it will remove SNPs correlated with your variants-of-interest before running PCA. </li><li><code>FLASHPCA_EXCLUSION_REGIONS</code> (optional, default: data/exclusion<em>regions</em>hg19.txt): A path to the flashpca special exclusion regions.</li><li><code>MAF_THRESHOLD</code> (optional, default: 0.01): Only variants with that minor allele frequency are considered</li><li><code>NB_PCS</code> (optional, default: 6): The number of PCA components to extract.</li></ul><h2 id="[Study-Designs](@ref)"><a class="docs-heading-anchor" href="#[Study-Designs](@ref)"><a href="../study_designs/#Study-Designs">Study Designs</a></a><a id="[Study-Designs](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Study-Designs](@ref)" title="Permalink"></a></h2><ul><li><strong><code>STUDY_DESIGN</code></strong> (required, default: &quot;FROM_PARAM_FILE&quot;): One of &quot;FROM_PARAM_FILE&quot;, &quot;FROM_ACTORS&quot;.</li></ul><p>If <code>STUDY_DESIGN</code>=<code>CUSTOM</code>:</p><ul><li><code>ESTIMANDS_FILE</code> (required): Path expression to the estimands file.</li></ul><p>If <code>STUDY_DESIGN</code>=<code>ALLELE_INDEPENDENT</code>:</p><ul><li><code>ESTIMANDS_FILE</code> (required): YAML configuration file.</li></ul><p>If <code>STUDY_DESIGN</code>=<code>FROM_ACTORS</code>:</p><ul><li><strong><code>BQTLS</code> (required)</strong>: A CSV file containing binding quantitative trait loci (bQTLs). If multiple transcription factors (TFs) are included in a single run, you must include a column called <code>TF</code>, which specifies the TF associated with each bQTL.</li><li><strong><code>TRANS_ACTORS</code> (required)</strong>: A prefix to CSV files containing quantitative trait loci potentially interacting with the previous bqtls. If multiple transcription factors (TFs) are included in a single run, you must include a column called <code>TF</code>, which specifies the TF associated with each transactor.</li><li><code>EXTRA_CONFOUNDERS</code> (optional, default: nothing): Path to additional confounders file, one per line, no header.</li><li><code>EXTRA_COVARIATES</code> (optional, default: nothing): Path to additional covariates file, one per line, no header.</li><li><code>ENVIRONMENTALS</code> (optional, default: nothing): Path to additional environmental treatments file, one per line, no header.</li><li><code>ORDERS</code> (optional, default: &quot;1,2&quot;): Comma separated list describing the order of desired interaction estimands, 1 for the ATE (no interaction), 2 for pairwise interactions etc... e.g. &quot;1,2&quot;</li></ul><h2 id="[Specifying-a-Targeted-Estimator](@ref)"><a class="docs-heading-anchor" href="#[Specifying-a-Targeted-Estimator](@ref)"><a href="../tmle/#Specifying-a-Targeted-Estimator">Specifying a Targeted Estimator</a></a><a id="[Specifying-a-Targeted-Estimator](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Specifying-a-Targeted-Estimator](@ref)" title="Permalink"></a></h2><ul><li><strong><code>ESTIMATOR_FILE</code> (required)</strong>: Julia configuration file describing the nuisance function learning algorithms.</li><li><code>POSITIVITY_CONSTRAINT</code> (optional, default: 0.01): Treatment variables rarest configuration should have at least that frequency.</li><li><code>KEEP_IC</code> (optional, default: false): For all parameters with an p-value below <code>PVAL_THRESHOLD</code>, the influence curve is saved.</li></ul><h2 id="[Correcting-for-population-relatedness](@ref)"><a class="docs-heading-anchor" href="#[Correcting-for-population-relatedness](@ref)"><a href="../sieve_variance/#Correcting-for-population-relatedness">Correcting for population relatedness</a></a><a id="[Correcting-for-population-relatedness](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Correcting-for-population-relatedness](@ref)" title="Permalink"></a></h2><ul><li><code>SVP</code> (optional, default: false): To perform Sieve Variance Correction.</li><li><code>GRM_NSPLITS</code> (optional, default: 100): To fasten GRM computation, it is typically split in batches.</li><li><code>NB_SVP_ESTIMATORS</code> (optional, default: 100): Number of sieve variance estimates per curve. Setting this value to 0 results in skipping sieve variance correction.</li><li><code>MAX_SVP_THRESHOLD</code> (optional, default: 0.9): Variance estimates are computed for tau ranging from 0 to MAX<em>SVP</em>THRESHOLD</li><li><code>PVAL_THRESHOLD</code> (optional, default: 0.05): To save computation time and disk, only parameters with a p-value below this threshold are considered for sieve variance correction.</li><li><code>SVP_ESTIMATOR_KEY</code> (optional, default: &quot;TMLE&quot;): The estimator to use for Sieve Variance Plateau correction.</li></ul><h2 id="[Tweaking-additional-behaviour](@ref)"><a class="docs-heading-anchor" href="#[Tweaking-additional-behaviour](@ref)"><a href="../miscellaneous/#Tweaking-additional-behaviour">Tweaking additional behaviour</a></a><a id="[Tweaking-additional-behaviour](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Tweaking-additional-behaviour](@ref)" title="Permalink"></a></h2><ul><li><code>CALL_THRESHOLD</code> (optional, default: 0.9): For putative causal variants (listed in the parameter files described in the <a href="../study_designs/#Study-Designs">Study Designs</a> section). If a individual&#39;s allele&#39;s probability is greater than the threshold, then it is called, otherwise it is considered missing.</li><li><code>BATCH_SIZE</code> (optional, default: 400): The set of parameters to be estimated is batched and the TMLE processes will run in parallel across batches on your platform.</li><li><code>OUTDIR</code> (optional, default: &quot;results&quot;): Output directory</li><li><code>RNG</code> (optional, default: 123): General random seed used where appropriate.</li></ul><h2 id="[Running-negative-control-checks](@ref)"><a class="docs-heading-anchor" href="#[Running-negative-control-checks](@ref)"><a href="../negative_control/#Running-negative-control-checks">Running negative control checks</a></a><a id="[Running-negative-control-checks](@ref)-1"></a><a class="docs-heading-anchor-permalink" href="#[Running-negative-control-checks](@ref)" title="Permalink"></a></h2><ul><li><code>MAX_PERMUTATION_TESTS</code> (optional, default: null): Arbitrarily limits the number of permutation tests performed.</li><li><code>ESTIMATOR_KEY</code> (optional, default: &quot;TMLE&quot;): The estimator from <code>ESTIMATOR_FILE</code> to use to asses significance.</li><li><code>PERMUTATION_ORDERS</code> (optional, default: &quot;1&quot;): A comma separating string defining the permutation test orders to be performed. (see <a href="../negative_control/#Permutation-tests">Permutation tests</a>)</li><li><code>MAF_MATCHING_RELTOL</code>(optional, default:0.05): Random variants are chosen with a similar MAF matched with the given relative tolerance.</li><li><code>N_RANDOM_VARIANTS</code>(optional, default: 10): For each hit, that many variants are randomly picked.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../negative_control/">« Running negative control checks</a><a class="docs-footer-nextpage" href="../runtime_considerations/">Some Runtime considerations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="auto">Automatic (OS)</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.2.1 on <span class="colophon-date" title="Tuesday 19 December 2023 15:06">Tuesday 19 December 2023</span>. Using Julia version 1.9.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
