<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Parameter specification · TarGene</title><script data-outdated-warner src="../assets/warner.js"></script><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/parameter_specification/"/><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">TarGene</a></span></div><form class="docs-search" action="../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Setting a data source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Confounding Adjustment</a></li><li class="is-active"><a class="tocitem" href>Parameter specification</a><ul class="internal"><li><a class="tocitem" href="#Parameter-Files"><span>Parameter Files</span></a></li><li><a class="tocitem" href="#Parameter-strategies"><span>Parameter strategies</span></a></li><li><a class="tocitem" href="#Parallelization"><span>Parallelization</span></a></li></ul></li><li><a class="tocitem" href="../tmle/">Targeted Maximum Likelihood Estimation</a></li><li><a class="tocitem" href="../sieve_variance/">Sieve Variance Plateau Correction</a></li><li><a class="tocitem" href="../nextflow_params/">Index of the pipeline parameters</a></li></ul></li><li><a class="tocitem" href="../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">User Guide</a></li><li class="is-active"><a href>Parameter specification</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Parameter specification</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/parameter_specification.md#" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Parameter-specification"><a class="docs-heading-anchor" href="#Parameter-specification">Parameter specification</a><a id="Parameter-specification-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-specification" title="Permalink"></a></h1><h2 id="Parameter-Files"><a class="docs-heading-anchor" href="#Parameter-Files">Parameter Files</a><a id="Parameter-Files-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-Files" title="Permalink"></a></h2><p>In this section, by parameter, we mean the statistical parameter that represents the scientific quantity of interest and will be estimated via TarGene. The complete specification of a parameter requires the description of a causal model which can be represented by the following graph.</p><div style="text-align:center">
<img src="assets/causal_graph.png" alt="Causal Model" style="width:400px;"/>
</div><p>Parameters are specified by a YAML file with one section for each variable in the causal model and a section for the parameters that need to be estimated from this causal model. Here is a succinct description for each section and the behaviour of the pipeline for each variable:</p><ul><li><code>Treatments</code>: The treatment variables, typically one or multiple SNPs and potential environmental exposures.</li><li><code>Confounders</code>: Confounding variables, typically the principal components which are computed by the pipeline. This section can (must) be ommited.</li><li><code>Covariates</code>: Additional covariates for the prediction of the traits. Not yet working and must be omitted.</li><li><code>Targets</code>: The traits of interest. The algorithm will loop through all traits in this section for the given treatments, confounders and covariates. This enables the reuse of the propensity score estimation. If this section is omitted (usually the case) all traits will be used.</li><li><code>Parameters</code>: For each target in <code>Targets</code> multiple parameters may be of interest depending on the exact case/control scenario of the treatment variables. For instance, since each genotyped locus can take up to 3 different values there could be up to 3 Average treatment Effect parameters if the treatment section consists of only one SNP.</li></ul><p>Since an example may be worth a thousand words, here are a couple of such files.</p><h3 id="Average-Treatment-Effect"><a class="docs-heading-anchor" href="#Average-Treatment-Effect">Average Treatment Effect</a><a id="Average-Treatment-Effect-1"></a><a class="docs-heading-anchor-permalink" href="#Average-Treatment-Effect" title="Permalink"></a></h3><p>One SNP <code>RSID_10</code> and 2 ATEs for all traits under study.</p><pre><code class="language-yaml hljs">Treatments:
  - RSID_10
Parameters:
  - name: GG_TO_AG
    RSID_10:
      control: GG
      case: AG
  - name: GG_TO_AA
    RSID_10:
      control: GG
      case: AA</code></pre><h3 id="Interaction-Average-Treatment-Effect"><a class="docs-heading-anchor" href="#Interaction-Average-Treatment-Effect">Interaction Average Treatment Effect</a><a id="Interaction-Average-Treatment-Effect-1"></a><a class="docs-heading-anchor-permalink" href="#Interaction-Average-Treatment-Effect" title="Permalink"></a></h3><p>Two interacting SNPs <code>RSID_10</code> and <code>RSID_100</code> with only one target <code>PHENOTYPE_1</code> and only one parameter.</p><pre><code class="language-yaml hljs">Treatments:
  - RSID_10
  - RSID_100

Targets:
  - PHENOTYPE_1

Parameters:
  - name: IATE
    RSID_10:
      control: GG
      case: AG
    RSID_100:
      control: GG
      case: AG</code></pre><h2 id="Parameter-strategies"><a class="docs-heading-anchor" href="#Parameter-strategies">Parameter strategies</a><a id="Parameter-strategies-1"></a><a class="docs-heading-anchor-permalink" href="#Parameter-strategies" title="Permalink"></a></h2><p>There are two main ways one can specify parameters that need to be estimated during a targene-pipeline run. This is done via the <code>MODE</code> parameter.</p><ul><li><code>MODE</code> = <code>GivenParameters</code></li></ul><p>In this setting, one typically write by &quot;hand&quot; or generate a set of parameter files as described above. The path to those parameters is then provided with the <code>PARAMETER_FILES</code> nextflow parameter.</p><ul><li><code>MODE</code> = <code>ASBxTransActors</code></li></ul><p>It is assumed that an initial set of variants has been pre-identified from a previous allele-specific binding (ASB) study as output by the <a href="https://git.ecdf.ed.ac.uk/oalmelid/baal-nf">ball-nf</a> pipeline: <code>ASB_FILES</code> parameter. It is also assumed that another set of potential trans-actors is given in a .csv file: <code>TRANS_ACTORS_FILE</code> parameter. In that scenario, the target parameter will be the Interaction Average Treatment Effect between every pair of SNPs. Additionally, if template parameters configuration files containing extra treatments are provided (via <code>PARAMETER_FILES</code>), nth-order interaction parameters will be generated.</p><h2 id="Parallelization"><a class="docs-heading-anchor" href="#Parallelization">Parallelization</a><a id="Parallelization-1"></a><a class="docs-heading-anchor-permalink" href="#Parallelization" title="Permalink"></a></h2><p>Since the same estimator for <code>p(T|W)</code> can be used for multiple target parameters, it may be useful to batch phenotypes using <code>PHENOTYPES_BATCH_SIZE</code>(default: 1) in order to reduce the computational burden.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../confounding_adjustment/">« Confounding Adjustment</a><a class="docs-footer-nextpage" href="../tmle/">Targeted Maximum Likelihood Estimation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.23 on <span class="colophon-date" title="Monday 26 September 2022 18:11">Monday 26 September 2022</span>. Using Julia version 1.8.1.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
