<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Index of Discovery Workflow Parameters · TarGene</title><meta name="title" content="Index of Discovery Workflow Parameters · TarGene"/><meta property="og:title" content="Index of Discovery Workflow Parameters · TarGene"/><meta property="twitter:title" content="Index of Discovery Workflow Parameters · TarGene"/><meta name="description" content="Documentation for TarGene."/><meta property="og:description" content="Documentation for TarGene."/><meta property="twitter:description" content="Documentation for TarGene."/><meta property="og:url" content="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><meta property="twitter:url" content="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">TarGene</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Examples</span><ul><li><a class="tocitem" href="../../examples/setup/">Setup: Read First</a></li><li><a class="tocitem" href="../../examples/gwas/">GWAS</a></li><li><a class="tocitem" href="../../examples/phewas/">PheWAS</a></li><li><a class="tocitem" href="../../examples/interactions/">Interaction Study</a></li></ul></li><li><span class="tocitem">User Guide</span><ul><li><a class="tocitem" href="../../overview/">Nextflow Basics</a></li><li><input class="collapse-toggle" id="menuitem-3-2" type="checkbox" checked/><label class="tocitem" for="menuitem-3-2"><span class="docs-label">The Discovery Workflow</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../overview/">Overview</a></li><li><a class="tocitem" href="../data_sources/">Genetic and Traits Data</a></li><li><a class="tocitem" href="../confounding_adjustment/">PCA Adjustment</a></li><li><a class="tocitem" href="../study_designs/">Defining the Estimands of Interest</a></li><li><a class="tocitem" href="../tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../miscellaneous/">Final Tweaks</a></li><li class="is-active"><a class="tocitem" href>Index of Discovery Workflow Parameters</a><ul class="internal"><li><a class="tocitem" href="#Arguments"><span>Arguments</span></a></li><li><a class="tocitem" href="#Main-Options"><span>Main Options</span></a></li><li><a class="tocitem" href="#Secondary-Options"><span>Secondary Options</span></a></li></ul></li><li><a class="tocitem" href="../outputs/">Understanding TarGene&#39;s Outputs</a></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3-3" type="checkbox"/><label class="tocitem" for="menuitem-3-3"><span class="docs-label">The Simulation Workflows (Experimental)</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../simulations/overview/">Overview</a></li><li><a class="tocitem" href="../../simulations/null_simulation/">Null Simulation</a></li><li><a class="tocitem" href="../../simulations/realistic_simulation/">Realistic Simulation</a></li></ul></li></ul></li><li><span class="tocitem">Secondary Workflows</span><ul><li><a class="tocitem" href="../../secondary_workflows/pca/">The PCA Workflow</a></li><li><a class="tocitem" href="../../secondary_workflows/make_dataset/">The Make Dataset Workflow</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../../developer_guide/project_organization/">Project Organization</a></li><li><a class="tocitem" href="../../developer_guide/contribution_guide/">Contributing</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">User Guide</a></li><li><a class="is-disabled">The Discovery Workflow</a></li><li class="is-active"><a href>Index of Discovery Workflow Parameters</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Index of Discovery Workflow Parameters</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/targene/configuration.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-Discovery-Workflow-Parameters"><a class="docs-heading-anchor" href="#Index-of-Discovery-Workflow-Parameters">Index of Discovery Workflow Parameters</a><a id="Index-of-Discovery-Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-Discovery-Workflow-Parameters" title="Permalink"></a></h1><h2 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h2><ul><li><strong><code>ESTIMANDS_CONFIG</code></strong>: YAML configuration file describing the effect sizes of interest, see the <a href="../study_designs/#Defining-the-Estimands-of-Interest">Defining the Estimands of Interest</a> section.</li><li><strong><code>ESTIMATORS_CONFIG</code></strong>: Estimator name or Julia file containing the description of the Targeted Estimators to use. To be consistent it should match the argument provided to the previous TarGene run.</li><li><strong><code>BGEN_FILES</code></strong>: Path to imputed BGEN files (optional for a GWAS).</li><li><strong><code>BED_FILES</code></strong>: Path to PLINK BED files.</li><li><strong><code>TRAITS_DATASET</code></strong>: Path to a traits dataset. If you are running this for a non-UKBB cohort, your sample IDs must be specified in the first column of this CSV file, with the column name <code>SAMPLE_ID</code>.</li></ul><h2 id="Main-Options"><a class="docs-heading-anchor" href="#Main-Options">Main Options</a><a id="Main-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Main-Options" title="Permalink"></a></h2><ul><li><strong><code>BATCH_SIZE</code>, default: 400</strong>: The set of estimands to be estimated is batched and the Targeted Learning processes will run in parallel across batches. This is the main driver of computational speed on High Performance Computing Platforms.</li><li><strong><code>COHORT</code> (default: &quot;UKBB&quot;)</strong>: Current default for this is UKBB. If set to a value other than UKBB, this will not run UKBB-specific trait extraction.</li><li><strong><code>POSITIVITY_CONSTRAINT</code> (default: 0.01)</strong>: When the list of estimands is generated or validated. Treatment variables&#39; rarest configuration should have at least that frequency. For example if the treatment variables are two variants with minor allele A and T respectively. The rarest configuration will be (AA, TT) and should have a frequency of at least <code>POSITIVITY_CONSTRAINT</code>.</li><li><strong><code>NB_PCS</code> (default: 6)</strong>: The number of PCA components to extract.</li><li><strong><code>SVP</code> (default: false)</strong>: Whether Sieve Variance Plateau correction should be performed.</li><li><strong><code>FLASHPCA_EXCLUSION_REGIONS</code> (default: assets/exclusion<em>regions</em>hg19.txt)</strong>: A path to the flashpca special exclusion regions.</li><li><strong><code>MAF_THRESHOLD</code> (default: 0.01)</strong>: Only variants with that minor allele frequency are considered</li><li><strong><code>LD_BLOCKS</code></strong>: A path to pre-identified linkage disequilibrium blocks to be removed from the BED files. It is good practice to specify <code>LD_BLOCKS</code>, as it will remove SNPs correlated with your variants-of-interest before running PCA.</li></ul><h3 id="If-COHORTUKBB-(default)"><a class="docs-heading-anchor" href="#If-COHORTUKBB-(default)">If <code>COHORT=UKBB</code> (default)</a><a id="If-COHORTUKBB-(default)-1"></a><a class="docs-heading-anchor-permalink" href="#If-COHORTUKBB-(default)" title="Permalink"></a></h3><ul><li><strong><code>UKB_CONFIG</code> (default: assets/ukbconfig.yaml)</strong>: YAML configuration file describing which traits should be extracted and how the population should be subsetted.</li><li><strong><code>UKB_ENCODING_FILE</code></strong>: If the <code>TRAITS_DATASET</code> is encrypted, an encoding file must be provided.</li><li><strong><code>UKB_WITHDRAWAL_LIST</code></strong>: List of participants withdrawn from the study.</li><li><strong><code>QC_FILE</code></strong>: Genotyping quality control file from the UK-Biobank study.</li></ul><h3 id="If-SVPtrue"><a class="docs-heading-anchor" href="#If-SVPtrue">If <code>SVP=true</code></a><a id="If-SVPtrue-1"></a><a class="docs-heading-anchor-permalink" href="#If-SVPtrue" title="Permalink"></a></h3><ul><li><strong><code>GRM_NSPLITS</code>, default: 100</strong>: To fasten GRM computation, it is typically split in batches.</li><li><strong><code>NB_SVP_ESTIMATORS</code>, default: 100</strong>: Number of sieve variance estimates per curve. Setting this value to 0 results in skipping sieve variance correction.</li><li><strong><code>MAX_SVP_THRESHOLD</code>, default: 0.9</strong>: Variance estimates are computed for tau ranging from 0 to MAX<em>SVP</em>THRESHOLD</li><li><strong><code>PVAL_THRESHOLD</code>, default: 0.05</strong>: Only results with a p-value below this threshold are considered for Sieve Plateau Variance correction.</li><li><strong><code>ESTIMATOR_KEY</code>, default: 1</strong>: Identifies an estimator from <code>ESTIMATORS_CONFIG</code>. The p-value for <code>PVAL_THRESHOLD</code> is computed using the result from this estimator.</li></ul><h2 id="Secondary-Options"><a class="docs-heading-anchor" href="#Secondary-Options">Secondary Options</a><a id="Secondary-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Secondary-Options" title="Permalink"></a></h2><ul><li><strong><code>RNG</code>, default: 123</strong>: General random seed used to induce permutation.</li><li><strong><code>VERBOSITY</code>, default: 0</strong>: Verbosity level of the the Workflow&#39;s processes.</li><li><strong><code>TL_SAVE_EVERY</code>, default: BATCH_SIZE</strong>: During the estimation process, results are appended to the file in chunks to free memory.</li><li><strong><code>KEEP_IC</code>, default: SVP</strong>: To save the Influence Curves for each estimate. Depending on the size of your dataset, this can result in massive disk usage.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../miscellaneous/">« Final Tweaks</a><a class="docs-footer-nextpage" href="../outputs/">Understanding TarGene&#39;s Outputs »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.7.0 on <span class="colophon-date" title="Thursday 5 September 2024 14:04">Thursday 5 September 2024</span>. Using Julia version 1.10.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
