<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Configuration · TarGene</title><meta name="title" content="Configuration · TarGene"/><meta property="og:title" content="Configuration · TarGene"/><meta property="twitter:title" content="Configuration · TarGene"/><meta name="description" content="Documentation for TarGene."/><meta property="og:description" content="Documentation for TarGene."/><meta property="twitter:description" content="Documentation for TarGene."/><meta property="og:url" content="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><meta property="twitter:url" content="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><link rel="canonical" href="https://TARGENE.github.io/targene-pipeline/targene/configuration/"/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">TarGene</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../../overview/">Overview</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">The TarGene Workflow </span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../overview/">The TarGene Workflow</a></li><li><a class="tocitem" href="../data_sources/">Setting a Data Source</a></li><li><a class="tocitem" href="../confounding_adjustment/">Adjusting for confounders</a></li><li><a class="tocitem" href="../study_designs/">Study Designs</a></li><li><a class="tocitem" href="../tmle/">Specifying a Targeted Estimator</a></li><li><a class="tocitem" href="../sieve_variance/">Correcting for population relatedness</a></li><li><a class="tocitem" href="../miscellaneous/">Tweaking additional behaviour</a></li><li class="is-active"><a class="tocitem" href>Configuration</a><ul class="internal"><li><a class="tocitem" href="#Main-Outputs"><span>Main Outputs</span></a></li><li><a class="tocitem" href="#Arguments"><span>Arguments</span></a></li><li><a class="tocitem" href="#Main-Options"><span>Main Options</span></a></li><li><a class="tocitem" href="#Secondary-Options"><span>Secondary Options</span></a></li></ul></li><li><a class="tocitem" href="../runtime_considerations/">Some Runtime considerations</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-2-3" type="checkbox"/><label class="tocitem" for="menuitem-2-3"><span class="docs-label">Negative Control</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../negative_control/overview/">Negative Control Overview</a></li><li><a class="tocitem" href="../../negative_control/permutation_tests/">The Permutation Test Workflow</a></li><li><a class="tocitem" href="../../negative_control/randomized_tests/">The Randomized Variants Workflow</a></li></ul></li><li><a class="tocitem" href="../../secondary_workflows/pca/">The PCA Workflow</a></li><li><a class="tocitem" href="../../secondary_workflows/make_dataset/">The Make Dataset Workflow</a></li></ul></li><li><span class="tocitem">Developper Guide</span><ul><li><a class="tocitem" href="../../developper_guide/project_organization/">Project Organization</a></li><li><a class="tocitem" href="../../developper_guide/contribution_guide/">Contributing</a></li></ul></li><li><a class="tocitem" href="../../associated_softwares/">Associated Softwares</a></li><li><a class="tocitem" href="../../publications/">Related Publications</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li><a class="is-disabled">The TarGene Workflow </a></li><li class="is-active"><a href>Configuration</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Configuration</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/TARGENE/targene-pipeline/blob/main/docs/src/targene/configuration.md#" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Configuration"><a class="docs-heading-anchor" href="#Configuration">Configuration</a><a id="Configuration-1"></a><a class="docs-heading-anchor-permalink" href="#Configuration" title="Permalink"></a></h1><h2 id="Main-Outputs"><a class="docs-heading-anchor" href="#Main-Outputs">Main Outputs</a><a id="Main-Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Main-Outputs" title="Permalink"></a></h2><p>The workflow will produce the following main outputs in the output directory (<code>OUTDIR</code>, defaults to <code>results</code>):</p><ul><li>An HDF5 file containing estimation results (<code>HDF5_OUTPUT</code>, default: results.hdf5)</li><li>An optional JSON file containing estimation results (<code>JSON_OUTPUT</code>)</li><li>An optional <code>svp.hdf5</code> containing Sieve Variance Plateau corrected variance estimates.</li><li>A Quantile-Quantile summary plot: QQ.png</li></ul><h2 id="Arguments"><a class="docs-heading-anchor" href="#Arguments">Arguments</a><a id="Arguments-1"></a><a class="docs-heading-anchor-permalink" href="#Arguments" title="Permalink"></a></h2><ul><li><strong><code>ESTIMATOR_FILE</code></strong>: Estimator name or Julia file containing the description of the Targeted Estimators to use. To be consistent it should match the argument provided to the previous TarGene run.</li><li><strong><code>BGEN_FILES</code></strong>: Path to imputed BGEN files from which the variants in <code>VARIANTS_LIST</code> will be extracted.</li><li><strong><code>BED_FILES</code></strong>: Path expression to PLINK BED files.</li><li><strong><code>TRAITS_DATASET</code></strong>: Path to a traits dataset. If you are running this for a non-UKBB cohort, your sample IDs must be specified in the first column of this CSV file, with the column name <code>SAMPLE_ID</code>.</li></ul><h2 id="Main-Options"><a class="docs-heading-anchor" href="#Main-Options">Main Options</a><a id="Main-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Main-Options" title="Permalink"></a></h2><ul><li><strong><code>STUDY_DESIGN</code> (default: CUSTOM)</strong>: How genetic variants and effect sizes are specified. See <a href="../study_designs/#Study-Designs">the study design section</a>.</li><li><strong><code>COHORT</code> (default: &quot;UKBB&quot;)</strong>: Current default for this is UKBB. If set to a value other than UKBB, this will not run UKBB-specific trait extraction.</li><li><strong><code>POSITIVITY_CONSTRAINT</code> (default: 0.01)</strong>: When the list of estimands is generated or validated. Treatment variables&#39; rarest configuration should have at least that frequency. For example if the treatment variables are two variants with minor allele A and T respectively. The rarest configuration will be (AA, TT) and should have a frequency of at least <code>POSITIVITY_CONSTRAINT</code>.</li><li><strong><code>NB_PCS</code> (default: 6)</strong>: The number of PCA components to extract.</li><li><strong><code>SVP</code> (default: false)</strong>: Whether Sieve Variance Plateau correction should be performed.</li><li><strong><code>FLASHPCA_EXCLUSION_REGIONS</code> (default: assets/exclusion<em>regions</em>hg19.txt)</strong>: A path to the flashpca special exclusion regions.</li><li><strong><code>MAF_THRESHOLD</code> (default: 0.01)</strong>: Only variants with that minor allele frequency are considered</li><li><strong><code>LD_BLOCKS</code></strong>: A path to pre-identified linkage disequilibrium blocks to be removed from the BED files. It is good practice to specify <code>LD_BLOCKS</code>, as it will remove SNPs correlated with your variants-of-interest before running PCA.</li></ul><h3 id="If-COHORTUKBB-(default)"><a class="docs-heading-anchor" href="#If-COHORTUKBB-(default)">If <code>COHORT=UKBB</code> (default)</a><a id="If-COHORTUKBB-(default)-1"></a><a class="docs-heading-anchor-permalink" href="#If-COHORTUKBB-(default)" title="Permalink"></a></h3><ul><li><strong><code>UKB_CONFIG</code> (default: assets/ukbconfig.yaml)</strong>: YAML configuration file describing which traits should be extracted and how the population should be subsetted.</li><li><strong><code>UKB_ENCODING_FILE</code></strong>: If the <code>TRAITS_DATASET</code> is encrypted, an encoding file must be provided.</li><li><strong><code>UKB_WITHDRAWAL_LIST</code></strong>: List of participants withdrawn from the study.</li><li><strong><code>QC_FILE</code></strong>: Genotyping quality control file from the UK-Biobank study.</li></ul><h3 id="If-STUDY_DESIGNCUSTOM-(default)"><a class="docs-heading-anchor" href="#If-STUDY_DESIGNCUSTOM-(default)">If <code>STUDY_DESIGN=CUSTOM</code> (default)</a><a id="If-STUDY_DESIGNCUSTOM-(default)-1"></a><a class="docs-heading-anchor-permalink" href="#If-STUDY_DESIGNCUSTOM-(default)" title="Permalink"></a></h3><ul><li><strong><code>ESTIMANDS_FILE</code></strong>: YAML configuration file describing the effect sizes of interest, see the <a href="../study_designs/#CUSTOM">custom study design section</a></li></ul><h3 id="If-STUDY_DESIGNALLELE_INDEPENDENT"><a class="docs-heading-anchor" href="#If-STUDY_DESIGNALLELE_INDEPENDENT">If <code>STUDY_DESIGN=ALLELE_INDEPENDENT</code></a><a id="If-STUDY_DESIGNALLELE_INDEPENDENT-1"></a><a class="docs-heading-anchor-permalink" href="#If-STUDY_DESIGNALLELE_INDEPENDENT" title="Permalink"></a></h3><ul><li><strong><code>ESTIMANDS_FILE</code></strong>: YAML configuration file describing the effect sizes of interest, see the <a href="../study_designs/#ALLELE_INDEPENDENT">allele independent section</a>.</li></ul><h3 id="If-STUDY_DESIGNFROM_ACTORS"><a class="docs-heading-anchor" href="#If-STUDY_DESIGNFROM_ACTORS">If <code>STUDY_DESIGN=FROM_ACTORS</code></a><a id="If-STUDY_DESIGNFROM_ACTORS-1"></a><a class="docs-heading-anchor-permalink" href="#If-STUDY_DESIGNFROM_ACTORS" title="Permalink"></a></h3><ul><li><strong><code>BQTLS</code></strong>: A CSV file containing binding quantitative trait loci (bQTLs). If multiple transcription factors (TFs) are included in a single run, you must include a column called <code>TF</code>, which specifies the TF associated with each bQTL.</li><li><strong><code>TRANS_ACTORS</code></strong>: A prefix to CSV files containing quantitative trait loci potentially interacting with the previous bqtls. If multiple transcription factors (TFs) are included in a single run, you must include a column called <code>TF</code>, which specifies the TF associated with each transactor.</li><li><strong><code>EXTRA_CONFOUNDERS</code>, default: nothing</strong>: Path to additional confounders file, one per line, no header.</li><li><strong><code>EXTRA_COVARIATES</code>, default: nothing</strong>: Path to additional covariates file, one per line, no header.</li><li><strong><code>ENVIRONMENTALS</code>, default: nothing</strong>: Path to additional environmental treatments file, one per line, no header.</li><li><strong><code>ORDERS</code>, default: &quot;1,2&quot;</strong>: Comma separated list describing the order of desired interaction estimands, 1 for the ATE (no interaction), 2 for pairwise interactions etc... e.g. &quot;1,2&quot;</li></ul><h3 id="If-SVPtrue"><a class="docs-heading-anchor" href="#If-SVPtrue">If <code>SVP=true</code></a><a id="If-SVPtrue-1"></a><a class="docs-heading-anchor-permalink" href="#If-SVPtrue" title="Permalink"></a></h3><ul><li><strong><code>GRM_NSPLITS</code>, default: 100</strong>: To fasten GRM computation, it is typically split in batches.</li><li><strong><code>NB_SVP_ESTIMATORS</code>, default: 100</strong>: Number of sieve variance estimates per curve. Setting this value to 0 results in skipping sieve variance correction.</li><li><strong><code>MAX_SVP_THRESHOLD</code>, default: 0.9</strong>: Variance estimates are computed for tau ranging from 0 to MAX<em>SVP</em>THRESHOLD</li><li><strong><code>PVAL_THRESHOLD</code>, default: 0.05</strong>: Only results with a p-value below this threshold are considered for Sieve Plateau Variance correction.</li><li><strong><code>ESTIMATOR_KEY</code>, default: TMLE</strong>: The p-value for <code>PVAL_THRESHOLD</code> is computed using the result from this estimator.</li></ul><h2 id="Secondary-Options"><a class="docs-heading-anchor" href="#Secondary-Options">Secondary Options</a><a id="Secondary-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Secondary-Options" title="Permalink"></a></h2><ul><li><strong><code>RNG</code>, default: 123</strong>: General random seed used to induce permutation.</li><li><strong><code>VERBOSITY</code>, default: 0</strong>: Verbosity level of the the Workflow&#39;s processes.</li><li><strong><code>BATCH_SIZE</code>, default: 400</strong>: The set of estimands to be estimated is batched and the Targeted Learning processes will run in parallel across batches.</li><li><strong><code>TL_SAVE_EVERY</code>, default: 50</strong>: During the estimation process, results are appended to the file in chunks to free memory.</li><li><strong><code>KEEP_IC</code>, default: SVP</strong>: To save the Influence Curves for each estimate. Depending on the size of your dataset, this can result in massive disk usage.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../miscellaneous/">« Tweaking additional behaviour</a><a class="docs-footer-nextpage" href="../runtime_considerations/">Some Runtime considerations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.3.0 on <span class="colophon-date" title="Monday 18 March 2024 15:04">Monday 18 March 2024</span>. Using Julia version 1.10.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
